src/s3ql/backends/b2/b2_backend.py

```diff
@@ -0,0 +1,807 @@
+            'isUploading': False
+        }
+
+    def _invalidate_upload_url(self, upload_url_info):
+        '''Removes a no longer working upload url from the available
+        upload urls.
+        '''
+
+        log.debug('invalidating upload url: %s %s', upload_url_info['hostname'], upload_url_info['path'])
+
+        if upload_url_info['connection'] is not None:
+            upload_url_info['connection'].disconnect()
+
+        try:
+            self.available_upload_url_infos.remove(upload_url_info)
+        except ValueError:
+            pass
+
+        log.debug('upload urls left: %d', len(self.available_upload_url_infos))
+
+    @staticmethod
+    def _b2_url_encode(s):
+        '''URL-encodes a unicode string to be sent to B2 in an HTTP header.'''
+
+        s = str(s)
+        encoded_s = urllib.parse.quote(s.encode('utf-8'), safe='/\\')
+        encoded_s = encoded_s.replace('\\', '=5C')
+
+        return encoded_s
+
+    @staticmethod
+    def _b2_url_decode(s, decode_plus=True):
+        '''Decodes a Unicode string returned from B2 in an HTTP header.
+
+        Returns a Python unicode string.
+        '''
+
+        # Use str() to make sure that the input to unquote is a str, not
+        # unicode, which ensures that the result is a str, which allows
+        # the decoding to work properly.
+
+        s = str(s)
+        decoded_s = s.replace('=5C', '\\')
+        if decode_plus:
+            decoded_s = urllib.parse.unquote_plus(decoded_s)
+        else:
+            decoded_s = urllib.parse.unquote(decoded_s)
+        return decoded_s
+
+    def _create_metadata_dict(self, metadata, chunksize=2048):
+        metadata_dict = {}
+
+        info_header_count = 0
+        for key in metadata.keys():
+            if not isinstance(key, str):
+                raise ValueError('dict keys must be str, not %s' % type(key))
+
+            value = metadata[key]
+            if (not isinstance(value, (str, bytes, int, float, complex, bool))
+                and value is not None):
+                raise ValueError('value for key %s (%s) is not elementary' % (key, value))
+
+            if isinstance(value, (bytes, bytearray)):
+                value = base64.b64encode(value)
+
+            buffer = ('%s: %s' % (repr(key), repr(value)))
+            buffer = urllib.parse.quote(buffer, safe='!@#$^*()=+/?-_\'"><\\| `.,;:~')
+
+            if len(buffer) < chunksize:
+                metadata_dict['meta-%03d' % info_header_count] = buffer
+                info_header_count += 1
+            else:
+                i = 0
+                while i * chunksize < len(buffer):
+                    k = 'meta-%03d' % info_header_count
+                    v = buffer[i * chunksize : (i + 1) * chunksize]
+                    metadata_dict[k] = v
+                    i += 1
+                    info_header_count += 1
+
+        # Backblaze B2 only allows 10 metadata headers (2 are used by format and md5)
+        assert info_header_count <= 8
```

Comment:
Since Backblaze can only handle 8 custom headers, we should probably be more efficient with them. This method will start a new header for every new metadata key. So we can only have at max 8 metadata keys. Works for now but it can easily break later down the line.

I think you should refactor this so that the metadata gets chunked into 2048 bytes chunks after you assembled the whole `repr`-buffer for the whole dict.

Reply:
I tried this back then but I did not succeed at that time (probably because I do not have a lot of python experience), the problem now is that I'm already using the backend with the metadata in this way since almost half a year.. and changing would need some conversion for my data..

Reply:
I just read that while there is the 10 info header limitation, Backblaze imposes no limit on the size of the headers. That would render the `chunksize` essentially unnecessary in this case. But the overall size of the headers can not exceed 7000 bytes. Though I don't know exactly what is counted towards that limit and/or we can count that.

Reply:
> I just read that while there is the 10 info header limitation, Backblaze imposes no limit on the size of the headers. That would render the `chunksize` essentially unnecessary in this case. [...]

It's probably safer to do chunking nevertheless – there could be crappy proxy servers in between S3QL and the B2 backend that do not like long header lines. But we could skip chunking for now and maybe introduce it when there is a user that needs chunking because of her outdated proxy infrastructure. (so add only one `X-Bz-Info-Meta-001` header – and when we decide to add chunking afterwards, that can be added in a backwards compatible manner).

> [...] But the overall size of the headers can not exceed 7000 bytes. [...]

I think 7000 bytes is quite enough. Here on my Swift Containers S3QL uses ~500 bytes of header data. Data objects store six metadata keys already (`encryption`, `signature`, `data`, `object_id`, `nonce` and `compression`) that and the two custom headers `Meta-Md5` and `Meta-Format` are eight custom headers that S3QL is using right now (of the 10 that are allowed).

> [...] Though I don't know exactly what is counted towards that limit and/or how we can count that.

from https://www.backblaze.com/b2/docs/files.html#httpHeaderSizeLimit
> HTTP Header Size Limit
The file name and file info must fit, along with the other necessary headers, within an 8KB limit imposed by some web servers and proxies. To ensure this, both now and in the future, B2 limits the combined header size for the file name and all file info to 7000 bytes. This limit applies to the fully encoded HTTP header line, including the carriage-return and newline. This header line is counted as 40 bytes:
> 
>    X-Bz-File-Name: %E8%87%AA%E7%94%B1.txt\r\n

(I am counting 41 bytes there, btw)

So `X-Bz-File-Name` and all `X-Bz-Info-Meta-*` headers count towards this 7000-bytes limit. S3QL might take something like 500 bytes of this limit for now. Leaves 6500 bytes for the file name – S3QL uses files names < 50 bytes so even with a very very long prefix we have a healthy safety margin.

Since it's not trivial to calculate the header size in advance and the limit is far bigger than what we need for now, I would skip adding an assertion for the 7000-byte limit. The B2 API will probably return a descriptive HTTP-Error anyways in this case.